docs/compatibility.md

```diff
@@ -0,0 +1,183 @@
+  + Forward compatibility: Add check so `--stablehlo-compatibility-check` emits a warning if `MyOp_v2` is used and cannot be downgraded.
+  + Backward compatibility: No change in deserialization of `MyOp`, libStablehlo@v2 can handle bytecode_v1.
+- 1/29/23: Remove forward compatibility warning 
+- 6/8/23: Remove `MyOp` in libStablehlo@v4
+  + Backward compatibility window over.
+  + Optionally rename `MyOp_v2` to `MyOp` via Migration Protocol
+
+### Changes in Dependent Dialects
+StableHLO will use forks of supported operations in dependent dialects for serialization / deserialization only. These forked operations allow StableHLO a finer control of the compatibility changes of operations in other dialects using the protocols above.
+
+_Proposal:_
+- Have stable shims of supported operations in other dialects.
+- Downgrade IR to shim before serializing `arith.addi → stablehlo.arith.addi`
+- Upgrade IR to other dialect on deserialization.
+
+With the above rules, we are able to provide compatibility guarantees for Migration and Additions. 
+
+_Example:_ Move `builtin.func → func.func`.
+
+![Dependent Dialect Change Protocol Diagram](images/dependent_dialect_change.png)
+
+There are still some open questions that require more discussion, regarding semantic changes or breaking changes to ops in other dialects. One option would be to do a deeper copy of the op into StableHLO and support it in our interpreter and transformations.
+
+## Compatibility Testing
+
+### Detecting Backward Incompatibility
+An approach similar to [LLVM bitcode compatibility testing](https://llvm.org/docs/DeveloperPolicy.html#ir-backwards-compatibility) can be used to detect backwards incompatibilities in StableHLO deserialization:
+> “After each X.Y release, compatibility.ll must be copied to compatibility-X.Y.ll. The corresponding bitcode file should be assembled using the X.Y build and committed as compatibility-X.Y.ll.bc.”
+
+There are large FileCheck integration tests for StableHLO (`ops_stablehlo_roundtrip.mlir`) and CHLO (`ops_chlo_roundtrip.mlir`). Simple breaking changes in backwards compatibility can be statically detected locally and in the “Build and Test” GitHub Action using the following procedure:
+
+- Create versions of the roundtrip files at certain increments.
+  + Create a copy as `ops_stablehlo_roundtrip_2022_09_06.mlir` and serialized `ops_stablehlo_roundtrip_2022_09_06_bc.mlir` for version created on Sep 6, 2022. 
+  + Copies of this file older than 6 months can be removed.
+  + The creation of these files should be scripted for easy testing, and we can automate this with GH actions as well.
+- In the most recent version of libStablehlo, deserialize all bytecode files from previous revisions.
+- Any failures to deserialize indicate a backwards incompatibility.
+
+This method has a bit of churn, we could automate the update of these files at a regular cadence to take the burden off the developer. Automate an update of the bytecode file every week.
```

Comment:
I suspect we can even automate a presubmit that there is an existing `ops_stablehlo_roundtrip_xxxxx.mlirbc` if the current codebase generates a different one than the latest existing.